main/src/ca/uwaterloo/flix/language/phase/unification/FastSetUnification.scala

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright 2024 Jonathan Lindegaard Starup
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.unification
+
+import scala.annotation.nowarn
+import scala.collection.immutable.SortedSet
+
+object FastSetUnification {
+
+  /**
+    * A common super-type for set formulas `f`, like `x1 ∩ x2 ∪ (e4 ∪ !c17)`.
+    *
+    * A set formula can contain elements, constants (fixed unknown sets), variables (unknown sets),
+    * complements, unions, and intersections.
+    *
+    * Variables and constants are collectively referred to as unknowns.
+    * A formula without unknowns is called ground.
+    *
+    * A set without unknowns can be evaluated to get a finite or a co-finite set of elements. The
+    * universe of elements is infinite so no finite set is equivalent to universe.
+    *
+    * Invariant: [[SetFormula.ElemSet]], [[SetFormula.Cst]], and [[SetFormula.Var]] must use
+    * disjoint integers (see [[unknowns]]).
+    */
+  sealed trait SetFormula {
+
+    import SetFormula.*
+
+    /** The set of [[Var]] in `this`. */
+    final def variables: SortedSet[Int] = this match {
+      case Univ => SortedSet.empty
+      case Empty => SortedSet.empty
+      case Cst(_) => SortedSet.empty
+      case Var(x) => SortedSet(x)
+      case ElemSet(_) => SortedSet.empty
+      case Compl(t) => t.variables
+      case Inter(_, _, varsPos, _, _, varsNeg, other) =>
+        SortedSet.from(
```

Comment:
Whats your reason for using iterators? As opposed to either folding or just ++?

Reply:
My worry is that -- some refactorings later -- iterators can be error-prone. And that the performance gain is not worth it?

Reply:
I don't know if the performance is noticeable, but I don't think it hurts here because the iterator is immediately created and given to `SortedSet`. it doesn't seem error prone

Reply:
My question is this: Why not just write a standard recursive function? That's what we do elsewhere. In only one place to we use iterators (SemanticTokens) and there we do it because benchmarking shows its critical for performance (the lists are very long).

Reply:
i like it because
- faster: this is a bottle neck of performance so everything counts
- low maintenance cost: its selfcontained
- better errors: its tailrecursive, so it allows us to give TooComplex more often instead of StackOverflow
- low effort: its already written

If anything is wrong/weird its easy to go back to recursive def. I think this is better